File: apps/worker/src/lib/storage.ts
import { Client } from "minio";
import { loadEnv } from "@folio/config";

let client: Client | null = null;

export function storageClient(): Client {
  if (client) return client;
  const env = loadEnv();
  client = new Client({
    endPoint: env.MINIO_ENDPOINT,
    port: env.MINIO_PORT,
    useSSL: Boolean(env.MINIO_USE_SSL),
    accessKey: env.MINIO_ACCESS_KEY,
    secretKey: env.MINIO_SECRET_KEY,
    region: env.MINIO_REGION,
  });
  return client;
}

export async function getObjectBuffer(key: string): Promise<Buffer> {
  const env = loadEnv();
  const stream = await storageClient().getObject(env.MINIO_BUCKET, key);
  const chunks: Buffer[] = [];
  for await (const chunk of stream) {
    chunks.push(Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk));
  }
  return Buffer.concat(chunks);
}

export async function putObject(key: string, body: Buffer, mimeType: string) {
  const env = loadEnv();
  await storageClient().putObject(env.MINIO_BUCKET, key, body, body.length, {
    "Content-Type": mimeType,
  });
}

export async function deleteObject(key: string) {
  const env = loadEnv();
  await storageClient().removeObject(env.MINIO_BUCKET, key);
}

File: apps/fastapi/app/storage.py
from __future__ import annotations

from datetime import datetime, timezone
from io import BytesIO
from uuid import uuid4

from minio import Minio

from .config import get_settings

_client: Minio | None = None


def storage_client() -> Minio:
    global _client
    if _client is None:
        settings = get_settings()
        _client = Minio(
            f"{settings.MINIO_ENDPOINT}:{settings.MINIO_PORT}",
            access_key=settings.MINIO_ACCESS_KEY,
            secret_key=settings.MINIO_SECRET_KEY,
            secure=settings.MINIO_USE_SSL,
            region=settings.MINIO_REGION,
        )
    return _client


def ensure_bucket() -> None:
    settings = get_settings()
    client = storage_client()
    if not client.bucket_exists(settings.MINIO_BUCKET):
        client.make_bucket(settings.MINIO_BUCKET, location=settings.MINIO_REGION)


def build_storage_key(owner_id: str, extension: str) -> str:
    now = datetime.now(timezone.utc)
    return f"documents/{owner_id}/{now.year}/{now.month:02d}/{uuid4()}{extension}"


def put_object(key: str, body: bytes, mime_type: str) -> None:
    settings = get_settings()
    storage_client().put_object(
        settings.MINIO_BUCKET,
        key,
        BytesIO(body),
        length=len(body),
        content_type=mime_type,
    )


def get_object_bytes(key: str) -> bytes:
    settings = get_settings()
    response = storage_client().get_object(settings.MINIO_BUCKET, key)
    try:
        return response.read()
    finally:
        response.close()
        response.release_conn()


def delete_object(key: str) -> None:
    settings = get_settings()
    storage_client().remove_object(settings.MINIO_BUCKET, key)


def bucket_exists() -> bool:
    settings = get_settings()
    return storage_client().bucket_exists(settings.MINIO_BUCKET)
